### PR TITLE
fs: Refactor `ZoneFile` for code reuse

### PR DESCRIPTION
**Pull request by metaspace**
Add a method to close the active zone.

### CODE REVIEW

**Comment by metaspace** [comment]
Pushed a series of commits to clean up locking around open/active zone management in `ZonedBlockDevice`.

**Comment by metaspace** [comment]
Pushed for new commit message wording.

File: fs/zbd_zenfs.cc [hunk 1]
```diff
@@ -342,12 +340,10 @@ IOStatus ZonedBlockDevice::Open(bool readonly, bool exclusive) {
   return IOStatus::OK();
 }
 
-void ZonedBlockDevice::NotifyIOZoneFull() {
-  active_io_zones_--;
-  zone_resources_.notify_one();
-}
+void ZonedBlockDevice::NotifyIOZoneFull() { active_io_zones_--; }
```
Comment (yhr) [inline fs/zbd_zenfs.cc:343]: Why did you remove the notify here?

Comment (metaspace) [reply fs/zbd_zenfs.cc:343]: It is as it says in the commit text. We don't use it.

**Comment by yhr** [comment]
The commit message was not very verbose.

I'd like to keep the notify_one() for future use, the zoned block device
should be notified upon zone state transitions - so this patch can just be
dropped.

**Comment by metaspace** [comment]
Ok I see. I will drop it then.

**Comment by metaspace** [comment]
> I'd like to keep the notify_one() for future use, the zoned block device
> should be notified upon zone state transitions - so this patch can just be
> dropped.

I see, I will drop that patch then.

**Comment by metaspace** [comment]
@yhr Addressed your comments and rebased on master.

File: fs/io_zenfs.cc [hunk 2]
```diff
@@ -229,15 +229,22 @@ ZoneFile::~ZoneFile() {
...
+}
+
+IOStatus ZoneFile::CloseActiveZone() {
+  IOStatus s = IOStatus::OK();
   if (active_zone_) {
-    s = active_zone_->CloseWR();
-    ReleaseActiveZone();
-    if (s.ok()) {
-      zbd_->NotifyIOZoneClosed();
+    active_zone_->CloseWR();
```
Comment (yhr) [inline fs/io_zenfs.cc:241]: Doh! You're not checking the return value

**Review by yhr — CHANGES_REQUESTED** [review]

File: fs/io_zenfs.cc (hunk 2 above)
Comment (metaspace) [reply fs/io_zenfs.cc:241]: oops

**Review by yhr — APPROVED** [review]
